### PR TITLE
migrate `sort` from the TH to Aten (CUDA)

### LINKED ISSUE #24637: Migrate `sort` from the TH to Aten (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

cc @heitorschueroff

### PR DESCRIPTION
**Pull request by zsef123**
Fixes #24637

## Changes
- Merging into just `sort` and `sort_out` both CPU and CUDA
- Replace `THCudaLongTensor_fillSliceWithIndex` to `_fill_indices`( and it move to `Sorting.cpp`)
- Replace `THCTensor_(copy)` in `THCTensor_(sort)`, In commit, handling in `sort_out`
- Changed TH functions to ATens
-- inlining functions like `nDimensionLegacyNoScalars`
-- Removes THCNumerics
-- Replace TH Macros (`THArgCheck`, ... ) to ATen Macros
- Changed macro in `sortViaThrust` to lambda

- Add Testing case about `sortViaThrust`
-- Changed `assertIsOrdered` to handling dynamic sizes of tensor

## Benchmark
- In V100 32GB, Driver : 450.80, cuda 10.1

<details><summary>Script</summary><p>

```import torch
from torch.utils.benchmark import Timer

torch.manual_seed(0)

experiments = (
    lambda: torch.rand(size=(10, 100), device="cuda"),
    lambda: torch.rand(size=(128, 128), device="cuda"),
    lambda: torch.rand(size=(128, 256), device="cuda"),
    lambda: torch.rand(size=(128, 512), device="cuda"),
    lambda: torch.rand(size=(128, 1024), device="cuda"),
    lambda: torch.rand(size=(512, 1024), device="cuda"),
    lambda: torch.rand(size=(1024, 1024), device="cuda"),

    lambda: torch.rand(size=(39, 222075), device="cuda"),
    lambda: torch.rand(size=(32, 262144), device="cuda"),
    lambda: torch.rand(size=(786842, 25), device="cuda"),
    lambda: torch.rand(size=(1048576, 16), device="cuda"),

    lambda: torch.rand(size=(384, 242, 4295), device="cuda"),
    lambda: torch.rand(size=(32, 29, 16, 27), device="cuda"),
    lambda: torch.rand(size=(12, 2, 6, 3, 11), device="cuda"),

)

for get_tensor in experiments:
    ndim = get_tensor().dim()
    for dim in range(ndim):
        for descending in [True, False]:
            x = get_tensor()
            timer = Timer(
                stmt="torch.sort(x, dim=dim, descending=descending)",
                globals={"x": x, "dim": dim, "descending": descending},
                label=f"dim:{dim}, size:{list(x.shape)}, descending:{descending}",
            )
            measurement = timer.blocked_autorange(min_run_time=5)
            print(f"{measurement.median * 1e6:>10.0f} us{'':>10}{measurement.label}")

```
</p>
</details>

<details><summary>This Commit</summary><p>

```
        37 us          dim:0, size:[10, 100], descending:True
        26 us          dim:0, size:[10, 100], descending:False
        26 us          dim:1, size:[10, 100], descending:True
        26 us          dim:1, size:[10, 100], descending:False
        26 us          dim:0, size:[128, 128], descending:True
        26 us          dim:0, size:[128, 128], descending:False
        26 us          dim:1, size:[128, 128], descending:True
        26 us          dim:1, size:[128, 128], descending:False
        26 us          dim:0, size:[128, 256], descending:True
        26 us          dim:0, size:[128, 256], descending:False
        29 us          dim:1, size:[128, 256], descending:True
        29 us          dim:1, size:[128, 256], descending:False
        26 us          dim:0, size:[128, 512], descending:True
        26 us          dim:0, size:[128, 512], descending:False
        32 us          dim:1, size:[128, 512], descending:True
        32 us          dim:1, size:[128, 512], descending:False
        37 us          dim:0, size:[128, 1024], descending:True
        37 us          dim:0, size:[128, 1024], descending:False
        40 us          dim:1, size:[128, 1024], descending:True
        40 us          dim:1, size:[128, 1024], descending:False
       203 us          dim:0, size:[512, 1024], descending:True
       203 us          dim:0, size:[512, 1024], descending:False
       118 us          dim:1, size:[512, 1024], descending:True
       118 us          dim:1, size:[512, 1024], descending:False
       311 us          dim:0, size:[1024, 1024], descending:True
       310 us          dim:0, size:[1024, 1024], descending:False
       213 us          dim:1, size:[1024, 1024], descending:True
       213 us          dim:1, size:[1024, 1024], descending:False
      3518 us          dim:0, size:[39, 222075], descending:True
      3517 us          dim:0, size:[39, 222075], descending:False
      8492 us          dim:1, size:[39, 222075], descending:True
      8493 us          dim:1, size:[39, 222075], descending:False
      3944 us          dim:0, size:[32, 262144], descending:True
      3980 us          dim:0, size:[32, 262144], descending:False
      7780 us          dim:1, size:[32, 262144], descending:True
      7782 us          dim:1, size:[32, 262144], descending:False
     24494 us          dim:0, size:[786842, 25], descending:True
     24499 us          dim:0, size:[786842, 25], descending:False
      3647 us          dim:1, size:[786842, 25], descending:True
      3656 us          dim:1, size:[786842, 25], descending:False
     19846 us          dim:0, size:[1048576, 16], descending:True
     19847 us          dim:0, size:[1048576, 16], descending:False
      4230 us          dim:1, size:[1048576, 16], descending:True
      4229 us          dim:1, size:[1048576, 16], descending:False
    226726 us          dim:0, size:[384, 242, 4295], descending:True
    226769 us          dim:0, size:[384, 242, 4295], descending:False
    221875 us          dim:1, size:[384, 242, 4295], descending:True
    222379 us          dim:1, size:[384, 242, 4295], descending:False
    470485 us          dim:2, size:[384, 242, 4295], descending:True
    470460 us          dim:2, size:[384, 242, 4295], descending:False
        83 us          dim:0, size:[32, 29, 16, 27], descending:True
        83 us          dim:0, size:[32, 29, 16, 27], descending:False
        89 us          dim:1, size:[32, 29, 16, 27], descending:True
        88 us          dim:1, size:[32, 29, 16, 27], descending:False
       116 us          dim:2, size:[32, 29, 16, 27], descending:True
       116 us          dim:2, size:[32, 29, 16, 27], descending:False
        77 us          dim:3, size:[32, 29, 16, 27], descending:True
        77 us          dim:3, size:[32, 29, 16, 27], descending:False
        27 us          dim:0, size:[12, 2, 6, 3, 11], descending:True
        27 us          dim:0, size:[12, 2, 6, 3, 11], descending:False
        26 us          dim:1, size:[12, 2, 6, 3, 11], descending:True
        27 us          dim:1, size:[12, 2, 6, 3, 11], descending:False
        26 us          dim:2, size:[12, 2, 6, 3, 11], descending:True
        26 us          dim:2, size:[12, 2, 6, 3, 11], descending:False
        27 us          dim:3, size:[12, 2, 6, 3, 11], descending:True
        27 us          dim:3, size:[12, 2, 6, 3, 11], descending:False
        27 us          dim:4, size:[12, 2, 6, 3, 11], descending:True
        26 us          dim:4, size:[12, 2, 6, 3, 11], descending:False
```
</p>
</details>

<details><summary>1.7.1</summary><p>

```
        26 us          dim:0, size:[10, 100], descending:True
        25 us          dim:0, size:[10, 100], descending:False
        25 us          dim:1, size:[10, 100], descending:True
        25 us          dim:1, size:[10, 100], descending:False
        26 us          dim:0, size:[128, 128], descending:True
        25 us          dim:0, size:[128, 128], descending:False
        25 us          dim:1, size:[128, 128], descending:True
        25 us          dim:1, size:[128, 128], descending:False
        25 us          dim:0, size:[128, 256], descending:True
        25 us          dim:0, size:[128, 256], descending:False
        29 us          dim:1, size:[128, 256], descending:True
        29 us          dim:1, size:[128, 256], descending:False
        26 us          dim:0, size:[128, 512], descending:True
        25 us          dim:0, size:[128, 512], descending:False
        32 us          dim:1, size:[128, 512], descending:True
        32 us          dim:1, size:[128, 512], descending:False
        37 us          dim:0, size:[128, 1024], descending:True
        37 us          dim:0, size:[128, 1024], descending:False
        40 us          dim:1, size:[128, 1024], descending:True
        40 us          dim:1, size:[128, 1024], descending:False
       204 us          dim:0, size:[512, 1024], descending:True
       203 us          dim:0, size:[512, 1024], descending:False
       118 us          dim:1, size:[512, 1024], descending:True
       117 us          dim:1, size:[512, 1024], descending:False
       310 us          dim:0, size:[1024, 1024], descending:True
       310 us          dim:0, size:[1024, 1024], descending:False
       213 us          dim:1, size:[1024, 1024], descending:True
       214 us          dim:1, size:[1024, 1024], descending:False
      3520 us          dim:0, size:[39, 222075], descending:True
      3534 us          dim:0, size:[39, 222075], descending:False
      8497 us          dim:1, size:[39, 222075], descending:True
      8495 us          dim:1, size:[39, 222075], descending:False
      3981 us          dim:0, size:[32, 262144], descending:True
      3983 us          dim:0, size:[32, 262144], descending:False
      7769 us          dim:1, size:[32, 262144], descending:True
      7772 us          dim:1, size:[32, 262144], descending:False
     24460 us          dim:0, size:[786842, 25], descending:True
     24463 us          dim:0, size:[786842, 25], descending:False
      3648 us          dim:1, size:[786842, 25], descending:True
      3658 us          dim:1, size:[786842, 25], descending:False
     19843 us          dim:0, size:[1048576, 16], descending:True
     19847 us          dim:0, size:[1048576, 16], descending:False
      4231 us          dim:1, size:[1048576, 16], descending:True
      4230 us          dim:1, size:[1048576, 16], descending:False
    226265 us          dim:0, size:[384, 242, 4295], descending:True
    226503 us          dim:0, size:[384, 242, 4295], descending:False
    221894 us          dim:1, size:[384, 242, 4295], descending:True
    222107 us          dim:1, size:[384, 242, 4295], descending:False
    469437 us          dim:2, size:[384, 242, 4295], descending:True
    469346 us          dim:2, size:[384, 242, 4295], descending:False
        83 us          dim:0, size:[32, 29, 16, 27], descending:True
        83 us          dim:0, size:[32, 29, 16, 27], descending:False
        89 us          dim:1, size:[32, 29, 16, 27], descending:True
        88 us          dim:1, size:[32, 29, 16, 27], descending:False
       116 us          dim:2, size:[32, 29, 16, 27], descending:True
       115 us          dim:2, size:[32, 29, 16, 27], descending:False
        77 us          dim:3, size:[32, 29, 16, 27], descending:True
        76 us          dim:3, size:[32, 29, 16, 27], descending:False
        26 us          dim:0, size:[12, 2, 6, 3, 11], descending:True
        26 us          dim:0, size:[12, 2, 6, 3, 11], descending:False
        26 us          dim:1, size:[12, 2, 6, 3, 11], descending:True
        26 us          dim:1, size:[12, 2, 6, 3, 11], descending:False
        47 us          dim:2, size:[12, 2, 6, 3, 11], descending:True
        26 us          dim:2, size:[12, 2, 6, 3, 11], descending:False
        26 us          dim:3, size:[12, 2, 6, 3, 11], descending:True
        27 us          dim:3, size:[12, 2, 6, 3, 11], descending:False
        26 us          dim:4, size:[12, 2, 6, 3, 11], descending:True
        26 us          dim:4, size:[12, 2, 6, 3, 11], descending:False
```
</p>
</details>

@VitalyFedyunin

### CODE REVIEW

**Comment by nikitaved** [comment]
Hey, @zsef123 ! Thank you very much for your work! I just skimmed through and for sorting with `thrust` you can adopt a `CompositeStridedAccessor` which was introduced in https://github.com/pytorch/pytorch/pull/39744.

**Comment by zsef123** [comment]
@nikitaved Thanks to commennts! I read that PR but I think I missed some works.

In #39744, the `CompositeRandomAccessor` working like `zip_iterator`, allows to handle 2 accessors.
And the `StridedRandomAccessor` works strided moves in pointer.
In SortKernel, `StridedRandomAccessor`s are created for each iteration controlled by `TensorIteratorConfig`.
That means sort tensors by each dims.

But the `sortViaThrust` sorting all dims by calling once thrust::sort. So `StridedRandomAccessor` seems not directly working on Vectorized Sort.

I tried to using `TensorIteratorConfig` same as CPU parts, Of course it is slow.(Also I changed `CompositeRandomAccessor` to working fine on CUDA.)

And I also have an idea to mix `thrust::for_each` and `sort`, but I don't think it's a good idea.

Have I ever missed something about this?

**Comment by zsef123** [comment]
Adopt a `CompositeStridedAccessor` in Thrust Kernel


Benchmarks
- Little slower than 1.7.1

<details><summary>Scripts</summary><p>

```
experiments = (
    lambda: torch.rand(size=(1, 2049), device="cuda"),
    lambda: torch.rand(size=(3, 2049), device="cuda"),
    lambda: torch.rand(size=(4, 2049), device="cuda"),
    lambda: torch.rand(size=(9, 2049), device="cuda"),
    lambda: torch.rand(size=(1, 12345), device="cuda"),
    lambda: torch.rand(size=(3, 12345), device="cuda"),
    lambda: torch.rand(size=(1, 23456), device="cuda"),
    lambda: torch.rand(size=(3, 23456), device="cuda"),
    lambda: torch.rand(size=(1, 2, 2049), device="cuda"),
    lambda: torch.rand(size=(2, 2, 2049), device="cuda"),
    lambda: torch.rand(size=(3, 4, 2049), device="cuda"),
)

for get_tensor in experiments:
    ndim = get_tensor().dim()
    for dim in range(ndim):
        if get_tensor().size(dim) < 2049:
            continue

        for descending in [True, False]:
            x = get_tensor()
            timer = Timer(
                stmt="torch.sort(x, dim=dim, descending=descending)",
                globals={"x": x, "dim": dim, "descending": descending},
                label=f"dim:{dim}, size:{list(x.shape)}, descending:{descending}",
            )
            measurement = timer.blocked_autorange(min_run_time=5)
            print(f"{measurement.median * 1e6:>10.0f} us{'':>10}{measurement.label}")
```
</p>
</details>

<details><summary>This commit</summary><p>

```
       134 us          dim:1, size:[1, 2049], descending:True
       134 us          dim:1, size:[1, 2049], descending:False
       156 us          dim:1, size:[3, 2049], descending:True
       156 us          dim:1, size:[3, 2049], descending:False
       165 us          dim:1, size:[4, 2049], descending:True
       166 us          dim:1, size:[4, 2049], descending:False
       185 us          dim:1, size:[9, 2049], descending:True
       185 us          dim:1, size:[9, 2049], descending:False
       176 us          dim:1, size:[1, 12345], descending:True
       191 us          dim:1, size:[1, 12345], descending:False
       233 us          dim:1, size:[3, 12345], descending:True
       236 us          dim:1, size:[3, 12345], descending:False
       214 us          dim:1, size:[1, 23456], descending:True
       220 us          dim:1, size:[1, 23456], descending:False
       256 us          dim:1, size:[3, 23456], descending:True
       259 us          dim:1, size:[3, 23456], descending:False
       170 us          dim:2, size:[1, 2, 2049], descending:True
       169 us          dim:2, size:[1, 2, 2049], descending:False
       184 us          dim:2, size:[2, 2, 2049], descending:True
       187 us          dim:2, size:[2, 2, 2049], descending:False
       209 us          dim:2, size:[3, 4, 2049], descending:True
       207 us          dim:2, size:[3, 4, 2049], descending:False
```
</p>
</details>

<details><summary>1.7.1</summary><p>

```
       103 us          dim:1, size:[1, 2049], descending:True
       103 us          dim:1, size:[1, 2049], descending:False
       136 us          dim:1, size:[3, 2049], descending:True
       133 us          dim:1, size:[3, 2049], descending:False
       142 us          dim:1, size:[4, 2049], descending:True
       142 us          dim:1, size:[4, 2049], descending:False
       159 us          dim:1, size:[9, 2049], descending:True
       158 us          dim:1, size:[9, 2049], descending:False
       152 us          dim:1, size:[1, 12345], descending:True
       152 us          dim:1, size:[1, 12345], descending:False
       184 us          dim:1, size:[3, 12345], descending:True
       182 us          dim:1, size:[3, 12345], descending:False
       169 us          dim:1, size:[1, 23456], descending:True
       167 us          dim:1, size:[1, 23456], descending:False
       204 us          dim:1, size:[3, 23456], descending:True
       205 us          dim:1, size:[3, 23456], descending:False
       129 us          dim:2, size:[1, 2, 2049], descending:True
       129 us          dim:2, size:[1, 2, 2049], descending:False
       142 us          dim:2, size:[2, 2, 2049], descending:True
       142 us          dim:2, size:[2, 2, 2049], descending:False
       159 us          dim:2, size:[3, 4, 2049], descending:True
       160 us          dim:2, size:[3, 4, 2049], descending:False
```
</p>
</details>

**Comment by nikitaved** [comment]
> @nikitaved Thanks to commennts! I read that PR but I think I missed some works.
> 
> In #39744, the `CompositeRandomAccessor` working like `zip_iterator`, allows to handle 2 accessors.
> And the `StridedRandomAccessor` works strided moves in pointer.
> In SortKernel, `StridedRandomAccessor`s are created for each iteration controlled by `TensorIteratorConfig`.
> That means sort tensors by each dims.

This means that TensorIterator iterates over each dimension but dimension `dim`, and one sorting is done per each such iteration over this dimension.

But I think you are right, with such an implementation parallelization is mostly done over the elements `prod([d in t.shape if d != dim])`.

> But the `sortViaThrust` sorting all dims by calling once thrust::sort. So `StridedRandomAccessor` seems not directly working on Vectorized Sort.
Could you show your benchmarks similar to what you did with the StridedRandomAccessor?

Thank you, @zsef123!

**Comment by zsef123** [comment]
Thanks to answer.

I already push the commit, apply CompositeRandomAccessor to the thrust kernel.

- remove make new tensor and copy parts
- using `StridedRandomAccessor` with `Stride 1`

I think I confused combine Comparator with Stride.

And in your answer, did you mean more benchmark datas? (https://github.com/pytorch/pytorch/pull/50887#issuecomment-778972700)

> > @nikitaved Thanks to commennts! I read that PR but I think I missed some works.
> > In #39744, the `CompositeRandomAccessor` working like `zip_iterator`, allows to handle 2 accessors.
> > And the `StridedRandomAccessor` works strided moves in pointer.
> > In SortKernel, `StridedRandomAccessor`s are created for each iteration controlled by `TensorIteratorConfig`.
> > That means sort tensors by each dims.
> 
> This means that TensorIterator iterates over each dimension but dimension `dim`, and one sorting is done per each such iteration over this dimension.
> 
> But I think you are right, with such an implementation parallelization is mostly done over the elements `prod([d in t.shape if d != dim])`.
> 
> > But the `sortViaThrust` sorting all dims by calling once thrust::sort. So `StridedRandomAccessor` seems not directly working on Vectorized Sort.
> 
> Could you show your benchmarks similar to what you did with the StridedRandomAccessor?
> 
> Thank you, @zsef123!

File: aten/src/ATen/native/CompositeRandomAccessor.h [hunk 1]
```diff
@@ -18,17 +23,44 @@ template <typename KeyAccessor, typename ValueAccessor>
 using CompositeRandomAccessorCPU =
   CompositeRandomAccessor<KeyAccessor, ValueAccessor, TupleInfoCPU>;
 
+#if defined(__CUDACC__)
+struct TupleInfoCUDA {
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/CompositeRandomAccessor.h:27]: It is better to use `aten/native/cuda/CompositeRandomAccessor.h` for sure. Yes, it was a mistake to leave CPU there, it should be CUDA instead.

**Comment by nikitaved** [comment]
> And in your answer, did you mean more benchmark datas? ([#50887 (comment)](https://github.com/pytorch/pytorch/pull/50887#issuecomment-778972700))

I mean, could you please provide your benchmarks prior to applying StridedAccessor?

**Comment by nikitaved** [comment]
>     * using `StridedRandomAccessor` with `Stride 1`
I am not sure you can do that for non-contiguous tensors, unless you make a contiguous copy.

**Comment by zsef123** [comment]
> > And in your answer, did you mean more benchmark datas? ([#50887 (comment)](https://github.com/pytorch/pytorch/pull/50887#issuecomment-778972700))
> 
> I mean, could you please run your benchmark tests on on commits prior to applying StridedAccessor?

At first commit https://github.com/pytorch/pytorch/pull/50887/commits/14277f441ef9d7606663d5dd5569d3984d250589, 

```118 us          dim:1, size:[1, 2049], descending:True
       118 us          dim:1, size:[1, 2049], descending:False
       138 us          dim:1, size:[3, 2049], descending:True
       138 us          dim:1, size:[3, 2049], descending:False
       146 us          dim:1, size:[4, 2049], descending:True
       146 us          dim:1, size:[4, 2049], descending:False
       160 us          dim:1, size:[9, 2049], descending:True
       162 us          dim:1, size:[9, 2049], descending:False
       154 us          dim:1, size:[1, 12345], descending:True
       153 us          dim:1, size:[1, 12345], descending:False
       185 us          dim:1, size:[3, 12345], descending:True
       184 us          dim:1, size:[3, 12345], descending:False
       169 us          dim:1, size:[1, 23456], descending:True
       169 us          dim:1, size:[1, 23456], descending:False
       207 us          dim:1, size:[3, 23456], descending:True
       207 us          dim:1, size:[3, 23456], descending:False
       132 us          dim:2, size:[1, 2, 2049], descending:True
       130 us          dim:2, size:[1, 2, 2049], descending:False
       146 us          dim:2, size:[2, 2, 2049], descending:True
       145 us          dim:2, size:[2, 2, 2049], descending:False
       163 us          dim:2, size:[3, 4, 2049], descending:True
       161 us          dim:2, size:[3, 4, 2049], descending:False
```

**Comment by nikitaved** [comment]
@zsef123 , then I apologize for wasting your time. Let's keep the fastest version then!

**Comment by zasdfgbnm** [comment]
What is the status of this PR? I am rewriting the thrust path with cub in https://github.com/pytorch/pytorch/pull/54626

**Comment by zasdfgbnm** [comment]
@zsef123 Are you still interested in working on this?

**Comment by zsef123** [comment]
> @zsef123 Are you still interested in working on this?

Sure I still interested but I just wait other reviews

**Comment by zsef123** [comment]
@zasdfgbnm @VitalyFedyunin @ngimel 
Integrate `stable` flags.

**Comment by ngimel** [comment]
@zsef123 #54626 will go in first as it is more limited, ditches thrust (that we need to do for other reasons) and significantly improves performance. I'll let you know when it's merged so you can rebase on that. 
Thanks for you work and sorry this PR has been hanging for so long, if your PR is not getting reviewed within couple weeks please feel free to ping assigned reviewers.

**Pull request closed without merging**